lib/charms/kratos/v0/kratos_endpoints.py

```diff
@@ -0,0 +1,156 @@
+# to 0 if you are raising the major API version
+LIBPATCH = 4
+
+RELATION_NAME = "kratos-endpoint-info"
+INTERFACE_NAME = "kratos_endpoints"
+logger = logging.getLogger(__name__)
+
+
+class KratosEndpointsRelationReadyEvent(EventBase):
+    """Event to notify the charm that the relation is ready."""
+
+
+class KratosEndpointsProviderEvents(ObjectEvents):
+    """Event descriptor for events raised by `KratosEndpointsProvider`."""
+
+    ready = EventSource(KratosEndpointsRelationReadyEvent)
+
+
+class KratosEndpointsProvider(Object):
+    """Provider side of the kratos-endpoint-info relation."""
+
+    on = KratosEndpointsProviderEvents()
+
+    def __init__(self, charm: CharmBase, relation_name: str = RELATION_NAME):
+        super().__init__(charm, relation_name)
+
+        self._charm = charm
+        self._relation_name = relation_name
+
+        events = self._charm.on[relation_name]
+        self.framework.observe(
+            events.relation_created, self._on_provider_endpoint_relation_created
+        )
+
+    def _on_provider_endpoint_relation_created(self, event: RelationCreatedEvent) -> None:
+        self.on.ready.emit()
+
+    def send_endpoint_relation_data(self, admin_endpoint: str, public_endpoint: str) -> None:
+        """Updates relation with endpoints info."""
+        if not self._charm.unit.is_leader():
+            return
+
+        relations = self.model.relations[self._relation_name]
+        endpoints_databag = {
+            "admin_endpoint": admin_endpoint,
+            "public_endpoint": public_endpoint,
+            "login_browser_endpoint": f"{public_endpoint}/self-service/login/browser",
+            "sessions_endpoint": f"{public_endpoint}/sessions/whoami",
+        }
+        for relation in relations:
+            relation.data[self._charm.app].update(endpoints_databag)
+
+
+class KratosEndpointsRelationError(Exception):
+    """Base class for the relation exceptions."""
+
+    pass
+
+
+class KratosEndpointsRelationMissingError(KratosEndpointsRelationError):
+    """Raised when the relation is missing."""
+
+    def __init__(self) -> None:
+        self.message = "Missing kratos-endpoint-info relation with kratos"
+        super().__init__(self.message)
+
+
+class KratosEndpointsRelationDataMissingError(KratosEndpointsRelationError):
+    """Raised when information is missing from the relation."""
+
+    def __init__(self, message: str) -> None:
+        self.message = message
+        super().__init__(self.message)
+
+
+class KratosEndpointsRequirer(Object):
+    """Requirer side of the kratos-endpoint-info relation."""
+
+    def __init__(self, charm: CharmBase, relation_name: str = RELATION_NAME):
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self.relation_name = relation_name
+
+    def get_kratos_endpoints(self) -> Optional[Dict]:
+        """Get the kratos endpoints."""
+        endpoints = self.model.relations[self.relation_name]
+        if len(endpoints) == 0:
```

Comment:
Maybe do `if not endpoints:`?

Reply:
This lib is imported from `kratos-operator`. We could do a little refactoring if you have more suggestions, though a pr needs to be opened in the other repository.